core/src/main/java/io/grpc/Metadata.java

```diff
@@ -279,22 +288,17 @@ public T apply(MetadataEntry metadataEntry) {
    */
   @Internal
   public byte[][] serialize() {
-    // One *2 for keys+values, one *2 to prevent resizing if a single key has multiple values
-    List<byte[]> serialized = new ArrayList<byte[]>(store.size() * 2 * 2);
-    for (Map.Entry<String, List<MetadataEntry>> keyEntry : store.entrySet()) {
-      for (int i = 0; i < keyEntry.getValue().size(); i++) {
-        MetadataEntry entry = keyEntry.getValue().get(i);
-        byte[] asciiName;
-        if (entry.key != null) {
-          asciiName = entry.key.asciiName();
-        } else {
-          asciiName = keyEntry.getKey().getBytes(US_ASCII);
-        }
-        serialized.add(asciiName);
-        serialized.add(entry.getSerialized());
+    // 2x for keys + values
+    byte[][] serialized = new byte[storeCount * 2][];
+    int i = 0;
+    for (Map.Entry<String, List<MetadataEntry>> storeEntry : store.entrySet()) {
+      for (MetadataEntry metadataEntry : storeEntry.getValue()) {
+        serialized[i++] = metadataEntry.key != null
+            ? metadataEntry.key.asciiName() : storeEntry.getKey().getBytes(US_ASCII);
+        serialized[i++] = metadataEntry.getSerialized();
       }
     }
-    return serialized.toArray(new byte[serialized.size()][]);
+    return serialized;
```

Comment:
I'd feel better if we asserted i == storeCount * 2 at the end, for better detection of bugs.